qutip/hsolve.py

```diff
@@ -0,0 +1,306 @@
+        current_pos = int(round(state2idx[tuple(nlabel)]))
+        Ltemp = sp.lil_matrix((Ntot, Ntot))
+        Ltemp[current_pos,current_pos] = 1
+        Ltemp.tocsr()
+        Lbig = Lbig + sp.kron(Ltemp,(-nlabel[0] * gam * spre(unit).data))
+        
+        for kcount in range(1,N):
+            Lbig = Lbig + sp.kron(Ltemp,(-nlabel[kcount] * (kcount * gj)
+            	   			 * spre(unit).data))
+        
+        for kcount in range(N):
+            if nlabel[kcount]>=1:
+            #find the position of the neighbour
+                nlabeltemp = copy(nlabel)
+                nlabel[kcount] = nlabel[kcount] -1
+                current_pos2 = int(round(state2idx[tuple(nlabel)]))
+                Ltemp = sp.lil_matrix(np.zeros((Ntot,Ntot)))
+                Ltemp[current_pos, current_pos2] = 1
+                Ltemp.tocsr()
+            # renormalized version:    
+                ci =  (4 * lam0 * gam * kb * Temperature * kcount
+                	   * gj/((kcount * gj)**2 - gam**2)) / (hbar**2)
+                if kcount==0:
+                    Lbig = Lbig + sp.kron(Ltemp,(-1j
+                    				 * (np.sqrt(nlabeltemp[kcount]
+                    				    / abs(c0)))
+                    				 * ((c0) * spre(Q).data
+                    		  	 	 - (np.conj(c0))
+                    				 * spost(Q).data)))
+                if kcount>0:                                                
+                    ci =  (4 * lam0 * gam * kb * Temperature * kcount
+                    	   * gj/((kcount * gj)**2 - gam**2)) / (hbar**2)
+                    Lbig = Lbig + sp.kron(Ltemp,(-1j
+                    				 * (np.sqrt(nlabeltemp[kcount]
+                    				    / abs(ci)))
+                    				 * ((ci) * spre(Q).data
+                    				 - (np.conj(ci))
+                    				 * spost(Q).data)))
+                nlabel = copy(nlabeltemp)
+
+        for kcount in range(N):
+            if ntotalcheck<=(Nc-1):
+                nlabeltemp = copy(nlabel)
+                nlabel[kcount] = nlabel[kcount] + 1
+                current_pos3 = int(round(state2idx[tuple(nlabel)]))
+            if current_pos3<=(Ntot):
+                Ltemp = sp.lil_matrix(np.zeros((Ntot,Ntot)))
+                Ltemp[current_pos, current_pos3] = 1
+                Ltemp.tocsr()
+            #renormalized   
+                if kcount==0:
+                    Lbig = Lbig + sp.kron(Ltemp,-1j
+                    			  * (np.sqrt((nlabeltemp[kcount]+1)
+                    			     * abs(c0)))
+                    			  * (spre(Q)- spost(Q)).data)
+                if kcount>0:
+                    ci = (4 * lam0 * gam * kb * Temperature * kcount
+                    	  * gj/((kcount * gj)**2 - gam**2)) / (hbar**2)
+                    Lbig = Lbig + sp.kron(Ltemp,-1j
+                    			  * (np.sqrt((nlabeltemp[kcount]+1) 
+                    			     * abs(ci)))
+                    			  * (spre(Q)- spost(Q)).data)
+            nlabel = copy(nlabeltemp)
```

Comment:
i'm guession that copy isnt needed here either, since nlabeltemp was assigned to a copy and it has not been modified